halo2_gadgets/src/poly_eval.rs

```diff
@@ -0,0 +1,484 @@
+//! A polynomial evaluation scheme
+//! Polynomial represents in coeffs, [c0, c1, c2, ...]
+//! P[x] = c0 * 1 + c1 * x + c2 * x^2 + ... cn * x^n
+//! where x is in a Fp
+
+use std::marker::PhantomData;
+
+use halo2_proofs::{
+    arithmetic::Field,
+    circuit::{AssignedCell, Chip, Layouter, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Selector},
+    poly::Rotation,
+};
+
+/// Global var for advice col layput
+const XINDEX: usize = 0;
+const COEFFINDEX: usize = 1;
+const ACCINDEX: usize = 2;
+// As end of the layout
+const EVALADVCOL: usize = ACCINDEX + 1;
+
+/// Polynomial represents in coeffs, [c0, c1, c2, ...]
+/// P[x] = c0 * 1 + c1 * x + c2 * x^2 + ... cn * x^n
+/// where selection of x and coeffs are from same finite field
+pub trait PolyEvalInstructions<F: Field>: Chip<F> {
+    /// Variable representing a coefficient.
+    /// Certainly it could be from another Field/Group,
+    /// orther than current one
```

Comment:
Oh this is unnecessary, here we have to restrict the coeff and elem in same field.
I will leave it for now ----- I am pretty sure reviewers might have more comments and solve them in one shot